lib/srv/reexec.go

```diff
@@ -740,63 +573,310 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+	homeDir := localUser.HomeDir
+	if !utils.IsDir(homeDir) {
+		homeDir = "/"
+	}
+	os.Setenv("HOME", localUser.HomeDir)
+	os.Setenv("USER", c.Login)
+
+	// Apply any additional environment variables from PAM.
+	for _, kv := range pamEnvironment {
+		key, value, ok := strings.Cut(strings.TrimSpace(kv), "=")
+		if !ok {
+			return errorWriter, teleport.RemoteCommandFailure, trace.BadParameter("bad environment variable from PAM, expected format \"key=value\" but got %q", kv)
+		}
+		if err := os.Setenv(key, value); err != nil {
+			return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err)
+		}
+	}
+
+	// Ensure that the working directory is one that the local user has access to.
+	if err := os.Chdir(homeDir); err != nil {
+		return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err, "failed to set working directory for networking process")
+	}
+
+	// Build request listener from first extra file that was passed to command.
 	ffd := os.NewFile(FirstExtraFile, "listener")
 	if ffd == nil {
 		return errorWriter, teleport.RemoteCommandFailure, trace.BadParameter("missing socket fd")
 	}
 
-	conn, err := uds.FromFile(ffd)
-	ffd.Close()
+	parentConn, err := uds.FromFile(ffd)
+	_ = ffd.Close()
 	if err != nil {
 		return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err)
 	}
 
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
+	writeErrorToConn := func(conn io.Writer, err error) {
+		conn.Write([]byte(err.Error()))
+	}
+
+	// Maintain a list of file paths to cleanup at the end of the process. This
+	// ensures that file cleanup is handled by the child in cases where the parent
+	// fails to cleanup due to filesystem namespace discrepancy (pam_namespace)
+	var filePathsToCleanup []string
+	defer func() {
+		for _, path := range filePathsToCleanup {
+			os.Remove(path)
+		}
+	}()
+
 	for {
-		buf := make([]byte, 1024)
+		// Wait for per-request connections to come from the parent process connection.
 		fbuf := make([]*os.File, 1)
-		n, fn, err := uds.ReadWithFDs(conn, buf, fbuf)
+		_, fn, err := uds.ReadWithFDs(parentConn, nil, fbuf)
 		if err != nil {
 			if utils.IsOKNetworkError(err) {
+				// parent connection closed, process should exit.
 				return errorWriter, teleport.RemoteCommandSuccess, nil
 			}
-			return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err)
+			writeErrorToConn(parentConn, trace.Wrap(err, "error reading networking request from parent"))
+			continue
 		}
-		addr := string(buf[:n])
+
 		if fn == 0 {
-			log.Errorf("Parent did not send a file descriptor for address %q.", addr)
+			writeErrorToConn(parentConn, trace.BadParameter("networking request requires a control file"))
+			continue
+		}
+
+		requestConn, err := uds.FromFile(fbuf[0])
+		_ = fbuf[0].Close()
+		if err != nil {
+			writeErrorToConn(parentConn, trace.Wrap(err, "failed to get a connection from control file"))
+			continue
+		}
+		context.AfterFunc(ctx, func() { requestConn.Close() })
```

Comment:
This will accumulate in the `context.cancelCtx` for every new request. The returned cleanup function must be called when we're done with the `requestConn` if this is used in a loop.